url85940d3d8e_ccp999_Expr4j_tgz-pJ8-org_boris_expr_util_ExprEvaluatorJ8/src/org/boris/expr/util/IO.java

```diff
@@ -0,0 +1,131 @@
+/*******************************************************************************
+ * This program and the accompanying materials
+ * are made available under the terms of the Common Public License v1.0
+ * which accompanies this distribution, and is available at 
+ * http://www.eclipse.org/legal/cpl-v10.html
+ * 
+ * Contributors:
+ *     Peter Smith
+ *******************************************************************************/
+package org.boris.expr.util;
+
+import java.io.BufferedReader;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileReader;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.io.OutputStream;
+import java.io.Reader;
+import java.io.StringWriter;
+import java.io.Writer;
+import java.net.URL;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.StringTokenizer;
+
+public class IO
+{
+    public static String toString(Reader r) throws IOException {
+        StringWriter sw = new StringWriter();
+        copy(r, sw, true);
+        return sw.toString();
+    }
+
+    public static String toString(InputStream is) throws IOException {
+        return toString(new InputStreamReader(is));
+    }
+
+    public static String toString(File f) throws IOException {
+        return toString(new FileReader(f));
+    }
+
+    public static void copy(Reader r, Writer w, boolean close)
+            throws IOException {
+        char[] buf = new char[4096];
+        int len = 0;
+        while ((len = r.read(buf)) > 0) {
+            w.write(buf, 0, len);
+        }
+        if (close) {
+            r.close();
+            w.close();
+        }
+    }
+
+    public static void copy(InputStream r, OutputStream w, boolean close)
+            throws IOException {
+        byte[] buf = new byte[4096];
+        int len = 0;
+        while ((len = r.read(buf)) > 0) {
+            w.write(buf, 0, len);
+        }
+        if (close) {
+            r.close();
+            w.close();
+        }
+    }
+
+    public static String getExtension(File f) {
+        if (f == null)
+            return null;
+        String n = f.getName();
+        int idx = n.lastIndexOf('.');
+        if (idx == -1) {
+            return null;
+        }
+        return n.substring(idx + 1);
+    }
+
+    public static String removeExtension(File f) {
+        if (f == null)
+            return null;
+        String n = f.getName();
+        int idx = n.lastIndexOf('.');
+        if (idx == -1)
+            return n;
+        return n.substring(0, idx);
+    }
+
+    public static byte[] toBytes(File file) throws IOException {
+        byte[] b = new byte[(int) file.length()];
+        FileInputStream fis = new FileInputStream(file);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *fis*. The resource is referenced at line 94. The resource is closed at line 95. There are other execution paths that do not contain closure statements, for example, when FileInputStream.read throws an exception. Ensure *fis* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).